### ISSUE TITLE
Window content does not resize on kwin "Quick Tile"

### ISSUE THREAD

**Issue by pjones123** [wez/wezterm#2063]
### What Operating System(s) are you seeing this problem on?

Linux X11

### WezTerm version

20220528-212940-84842aaa

### Did you try the latest nightly build to see if the issue is better (or worse!) than your current version?

Yes, and I updated the version box above to show the version of the nightly that I tried

### Describe the bug

When resizing the window with kwin keyboard shortcuts for "Quick Tile Window to the [direction]", the contents of the window do not change to reflect the new window size. The window border (shown by kwin) is visible, but the window is transparent in areas that do not overlap the previous position, but the contents remain unchanged in areas that overlap the previous position. The issue also sometimes occurs when launching the application, where the entire window contents are not visible until updated by mouse interaction. I have been unable to determine the conditions under which the issue occurs in this way, but it appears to happen roughly 50% of the time.

This means that using "Quick Tile Window to the Bottom" after it is correctly showing tiled to the left, the bottom right corner is transparent, and the bottom left corner shows the bottom half of the terminal.

The window immediately updates to its new position and size after movement/resize using the mouse, or by changing focus to another window and then back to the terminal.

### To Reproduce

Launch the application from KDE plasma (on X11, wayland untested) using kwin. Use configured keyboard shortcuts to invoke kwin's "Quick Tile Window to [direction]".

### Configuration

no config

### Expected Behavior

The contents of the window should immediately update to reflect the window's new size and/or position, as it does when changed by the mouse.

### Logs

21:05:12.732  INFO   wezterm_gui::termwindow > OpenGL initialized! NVIDIA GeForce GTX 960/PCIe/SSE2 4.6.0 NVIDIA 515.43.04 is_context_loss_possible=false wezterm version: 20220528-212940-84842aaa

### Anything else?

"Quick Tile Window to the Left" after launch:
![image](https://user-images.githubusercontent.com/64071912/171492567-63635eb4-467f-4a94-8fb0-56e9d540933d.png)
"Quick Tile Window to the Bottom" after being tiled to the left:
![image](https://user-images.githubusercontent.com/64071912/171492653-cb2758fe-2027-4381-ae26-57a0855f0a45.png)

**Comment by pjones123** [comment]
Here's the output from that. I've added some annotations to show what I did to cause the behaviour.
```
15:40:49.770  TRACE  window::os::x11::window > {"DVI-D-0": Rect(1920x1080 at (0, 0)), "HDMI-0": Rect(1920x1080 at (1920, 0))}
15:40:49.770  TRACE  window::os::x11::window > virtual: Rect(3840x1080 at (0, 0))
15:40:49.770  TRACE  window::os::x11::window > main: Rect(1920x1080 at (0, 0))
15:40:49.772  TRACE  window::os::x11::window > PropertyNotifyEvent WM_CLASS
15:40:49.772  TRACE  window::os::x11::window > PropertyNotifyEvent _NET_WM_PID
15:40:49.772  TRACE  window::os::x11::window > PropertyNotifyEvent WM_PROTOCOLS
15:40:49.772  TRACE  window::os::x11::window > PropertyNotifyEvent _MOTIF_WM_HINTS
15:40:49.772  TRACE  window::os::x11::window > PropertyNotifyEvent _KDE_NET_WM_USER_CREATION_TIME
15:40:49.773  TRACE  window::os::x11::window > PropertyNotifyEvent WM_NAME
15:40:49.773  TRACE  window::os::x11::window > PropertyNotifyEvent _NET_WM_NAME
15:40:49.773  TRACE  window::os::x11::window > PropertyNotifyEvent _NET_WM_ICON
15:40:49.914  INFO   wezterm_gui::termwindow > OpenGL initialized! NVIDIA GeForce GTX 960/PCIe/SSE2 4.6.0 NVIDIA 515.43.04 is_context_loss_possible=false wezterm version: 20220528-212940-84842aaa
15:40:49.916  TRACE  window::os::x11::window > About to paint, but we're unsure about focus; querying!
15:40:49.916  TRACE  window::os::x11::window > Do I have focus? true
15:40:49.916  TRACE  window::os::x11::window > About to paint, but we've unsure about geometry; querying!
15:40:49.917  TRACE  window::os::x11::window > geometry is 656x424 vs. our initial 656x424
15:40:49.944  TRACE  window::os::x11::window > PropertyNotifyEvent WM_NAME
15:40:49.944  TRACE  window::os::x11::window > PropertyNotifyEvent _NET_WM_NAME
15:40:50.006  TRACE  window::os::x11::window > PropertyNotifyEvent _NET_WM_ICON_GEOMETRY
-- quick tile left, behaviour matches that of the first screenshot above
15:40:55.297  TRACE  window::os::x11::window > Calling focus_change(false)
-- force update by using the mouse to move the window
15:41:07.635  TRACE  window::os::x11::window > CONFIGURE_NOTIFY: width 656 -> 912, height 424 -> 1018, dpi 96 -> 96
15:41:07.658  TRACE  window::os::x11::window > Ignoring CONFIGURE_NOTIFY (912x1018 dpi=96) because width,height,dpi are unchanged
-- quick tile down, behaviour matches that of the second screenshot above
15:41:14.772  TRACE  window::os::x11::window > Calling focus_change(true)
-- force update by using the mouse to move the window
15:41:20.286  TRACE  window::os::x11::window > CONFIGURE_NOTIFY: width 912 -> 1838, height 1018 -> 501, dpi 96 -> 96
-- close window with alt+f4
15:41:23.488  TRACE  window::os::x11::window > Calling focus_change(false)
15:41:23.490  TRACE  window::os::x11::window > Calling focus_change(true)
```

**Comment by pjones123** [comment]
Thanks for the quick response.

I built it with that new commit, but unfortunately it does not fix the issue. From the logs, it appears that this check is working in that it triggers an update of the window geometry despite the lack of `CONFIGURE_NOTIFY`. However, the `geometry is {}x{} vs. our initial {}x{}` indicates that the new geometry is that of the window *before* the tiling. This means that there is a noticeable difference in behaviour compared with before this commit, in that it is clear that the geometry is updated to the previous position, while before it was not being updated at all.

Perhaps I should also open an issue with kwin regarding the lack of `CONFIGURE_NOTIFY`?

**Comment by wez** [comment]
Hmm, so I'm assuming that the focus changed even was from you clicking away to perform the tile action.
I've pushed another commit that logs expose events and that tries to invalidate based on the newly exposed area being obviously large than what we thought.
I may add an option to always invalidate the geometry on an expose event: in class X11 that might suck, but with compositing window managers being the more normal thing these days, I think it might be OK.

> Perhaps I should also open an issue with kwin regarding the lack of `CONFIGURE_NOTIFY`?

Yes please!
I think it would be good for wezterm to have a workaround, but it would be better to fix this issue at its heart!

**Comment by pjones123** [comment]
As I have no had this issue with any other application, I decided to see whether the issue of the missing `CONFIGURE_NOTIFY` could be with wezterm or the x bindings that it uses before submitting an issue to kwin.

I wrote a simple program using the same version of xcb, and had it log every `CONFIGURE_NOTIFY` event. From my limited testing, I was unable to replicate the issue; `CONFIGURE_NOTIFY` was sent as expected every time I tiled the window.

Is it possible that wezterm is erroneously filtering out some of these events?

**Comment by pjones123** [comment]
The window ID seems consistent for all `CONFIGURE_NOTIFY` events.

I added some logging in `process_queued_xcb` to log all events. Is this the only place where events are received?

This showed that the only events sent during "quick tile" were `KeyPress` and `KeyRelease` for the keyboard shortcut, `FocusIn` and `FocusOut` (seemingly arbitrarily, and with no visual indication of a loss of focus). `ConfigureNotify` is sent as expected when the window is moved/resized by the mouse.

I'm hesitant to open an issue elsewhere as this problem occurs only with wezterm.

**Comment by wez** [comment]
I installed KDE to test this here and I couldn't reproduce it.  I'm running on AMD hardware.

Over in #1992, other nvidia users have reported that switching away from the nvidia drivers seems to help. I'd originally discounted that as wezterm's EGL usage occurs at a different layer from X11 messages, but perhaps there is something weird going on with the EGL/GLX stuff in the X server when EGL is enabled and that impacts whether it emits these events in all the right places?

I think the easiest thing to try to verify this is to set `front_end = "Software"` in your wezterm config. That will switch to using Mesa's CPU based rendering.

An alternative is to set `__EGL_VENDOR_LIBRARY_FILENAMES=/usr/share/glvnd/egl_vendor.d/50_mesa.json` in the environment before launching wezterm.  It should have a similar effect.

**Comment by pjones123** [comment]
That appears to be the issue - nvidia again!
Setting `front_end  = "Software"` didn't seem to have any effect (in behaviour or logs), but setting `__EGL_VENDOR_LIBRARY_FILENAMES=/usr/share/glvnd/egl_vendor.d/50_mesa.json` fixed the issue - the behaviour is as expected and the configure notify events are sent. This does also cause the "weird flicker" described in https://github.com/wez/wezterm/issues/1992#issuecomment-1140074833

Is this believed to be an issue with the nvidia drivers themselves? It seems unlikely to me given the number of programs that work correctly with EGL/nvidia.

**Comment by wez** [comment]
I just pushed a commit that enables the Present X extension and subscribes to its version of CONFIGURE_NOTIFY.
The thinking is that maybe the nvidia implementation is generating those but not the X level events of the same name.
Please give it a try and let me know if it is effective!

**Comment by wez** [comment]
I'd love to see the `WEZTERM_LOG=window::os::x11::window=trace,info` logs with this build!

**Comment by pjones123** [comment]
```
13:29:06.435  TRACE  window::os::x11::window > {"HDMI-0": Rect(1920x1080 at (1920, 0)), "DVI-D-0": Rect(1920x1080 at (0, 0))}
13:29:06.435  TRACE  window::os::x11::window > virtual: Rect(3840x1080 at (0, 0))
13:29:06.435  TRACE  window::os::x11::window > main: Rect(1920x1080 at (0, 0))
13:29:06.438  TRACE  window::os::x11::window > PropertyNotifyEvent _KDE_NET_WM_USER_CREATION_TIME
13:29:06.438  TRACE  window::os::x11::window > PropertyNotifyEvent WM_CLASS
13:29:06.438  TRACE  window::os::x11::window > PropertyNotifyEvent _NET_WM_PID
13:29:06.438  TRACE  window::os::x11::window > PropertyNotifyEvent WM_PROTOCOLS
13:29:06.438  TRACE  window::os::x11::window > PropertyNotifyEvent _MOTIF_WM_HINTS
13:29:06.440  TRACE  window::os::x11::window > PropertyNotifyEvent WM_NAME
13:29:06.441  TRACE  window::os::x11::window > PropertyNotifyEvent _NET_WM_NAME
13:29:06.654  INFO   wezterm_gui::termwindow > OpenGL initialized! NVIDIA GeForce GTX 960/PCIe/SSE2 4.6.0 NVIDIA 515.43.04 is_context_loss_possible=false wezterm version: 20220603-180851-da59a222
13:29:06.654  TRACE  window::os::x11::window > About to paint, but we're unsure about focus; querying!
13:29:06.654  TRACE  window::os::x11::window > Do I have focus? true
13:29:06.654  TRACE  window::os::x11::window > About to paint, but we've unsure about geometry; querying!
13:29:06.654  TRACE  window::os::x11::window > geometry is 656x424 vs. our initial 656x424
13:29:06.683  TRACE  window::os::x11::window > PropertyNotifyEvent WM_NAME
13:29:06.683  TRACE  window::os::x11::window > PropertyNotifyEvent _NET_WM_NAME
13:29:06.740  TRACE  window::os::x11::window > PropertyNotifyEvent _NET_WM_ICON_GEOMETRY
-- quick tile left, behaviour unchanged.
13:29:10.546  TRACE  window::os::x11::window > Calling focus_change(false)
13:29:10.547  TRACE  window::os::x11::window > About to paint, but we've unsure about geometry; querying!
-- queries window geometry (commit c3b1aa72d3927ced0a69b894348e05b721af409e) but receives the pre-tile geometry.
13:29:10.547  TRACE  window::os::x11::window > geometry is 656x424 vs. our initial 656x424
-- use mouse to trigger ConfigureNotify.
13:29:13.820  TRACE  window::os::x11::window > X::ConfigureNotify: width 656 -> 911, height 424 -> 1018, dpi 96 -> 96
-- quick tile down, behaviour unchanged.
13:29:19.633  TRACE  window::os::x11::window > Calling focus_change(true)
13:29:19.633  TRACE  window::os::x11::window > About to paint, but we've unsure about geometry; querying!
-- again, these are the dimensions of the window when it was tiled left.
13:29:19.633  TRACE  window::os::x11::window > geometry is 911x1018 vs. our initial 911x1018
-- use mouse to trigger ConfigureNotify.
13:29:24.075  TRACE  window::os::x11::window > X::ConfigureNotify: width 911 -> 1838, height 1018 -> 501, dpi 96 -> 96
-- close window.
13:29:30.358  TRACE  window::os::x11::window > Calling focus_change(false)
13:29:30.358  TRACE  window::os::x11::window > About to paint, but we've unsure about geometry; querying!
13:29:30.359  TRACE  window::os::x11::window > geometry is 1838x501 vs. our initial 1838x501
13:29:30.365  TRACE  window::os::x11::window > Calling focus_change(true)
```

You will notice the lack of `Present::ConfigureNotify` here, even when clicking the mouse to trigger an update. `Present::ConfigureNotify` is sent, as well as `X::ConfigureNotify` when resizing the window with the mouse, and sometimes when moving the window with the mouse.

These events also seem to be missing when the tiling is triggered by dragging the window the edge of the screen, although a lot less frequently than when using the keyboard shortcuts.

There is a strange lack of consistency and seeming randomness within both points here.

**Comment by wez** [comment]
Thanks! OK, I've pushed another attempt to work around this: 100ms after we receive the focus events, we'll invalidate and hope to pick up the new geometry.  Could you give that a try?  Hopefully 100ms is long enough, but you could try raising that to find out what threshold works.

**Comment by pjones123** [comment]
That delay fixes the issue - the delay is visible so is clearly working, and is definitely long enough. I will try to see how low this can go while remaining consistently effective, but I'd imagine that this value will vary between systems. Perhaps this should be added to config?

Obviously this isn't a nice workaround, but I can't imagine that much more can be done if the root cause is the nvidia drivers themselves, which seems quite likely now.

I won't bother contacting nvidia about this - I'm sure you're aware of how much they care about these things, especially on linux.

**Comment by pjones123** [comment]
I have noticed another problem occurring as a result of this issue. The value sent with the focus event that is being used to repaint the window is unreliable. It sends either `true` or `false`, seemingly arbitrarily. This means that sometimes the window is being rendered to not have focus, when it actually is focused.

I've pushed another commit to my PR with a fix for this, by assuming that the window still has focus if its size/position has changed during the event. I've tested it and it works, and I can't think of any situations where it could cause problems.
